amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/polling/PrefetchRecordsPublisher.java

```diff
@@ -224,12 +244,34 @@ public void cancel() {
     }
 
     @Override
-    public void notify(RecordsDeliveryAck ack) {
-        eventDeliveryLock.release();
+    public synchronized void notify(RecordsDeliveryAck recordsDeliveryAck) {
+        final RecordsRetrieved recordsToCheck = peekNextResult();
+        // Verify if the ack matches the head of the queue and evict it.
+        if (recordsToCheck != null && recordsToCheck.batchUniqueIdentifier()
+                .equals(recordsDeliveryAck.batchUniqueIdentifier())) {
+            evictNextResult();
+            // Upon evicting, check if queue is empty. if yes, then give the drain control back to publisher thread.
+            if(getRecordsResultQueue.isEmpty()) {
+                log.debug("{} : Publisher thread takes over the draining control. Queue Size : {}, Demand : {}", shardId,
+                        getRecordsResultQueue.size(), requestedResponses.get());
+                shouldDrainEventOnAck.set(false);
+            } else {
+                // Else attempt to drain the queue.
+                drainQueueForRequests();
+            }
+        } else {
+            // Log and ignore any other ack received. As long as an ack is received for head of the queue
+            // we are good. Any stale or future ack received can be ignored, though the latter is not feasible
+            // to happen.
```

Comment:
Another reason to come here is when the queue is reset before the ack comes back. Add this too in the comment?

Reply:
That would be considered as a stale ack.